ansible_base/rbac/models.py

```diff
@@ -0,0 +1,683 @@
+import logging
+
+# CRUM for getting the requesting user
+from crum import get_current_user
+
+# Django
+from django.conf import settings
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import connection, models
+from django.db.models.functions import Cast
+from django.utils.translation import gettext_lazy as _
+
+# Django-rest-framework
+from rest_framework.exceptions import ValidationError
+
+# ansible_base lib functions
+from ansible_base.lib.abstract_models.common import CommonModel
+
+# ansible_base RBAC logic imports
+from ansible_base.lib.utils.models import is_add_perm
+from ansible_base.rbac.permission_registry import permission_registry
+from ansible_base.rbac.prefetch import TypesPrefetch
+from ansible_base.rbac.validators import validate_assignment, validate_permissions_for_model
+
+logger = logging.getLogger('ansible_base.rbac.models')
+
+
+class DABPermission(models.Model):
+    "This is a minimal copy of auth.Permission for internal use"
+
+    name = models.CharField("name", max_length=255)
+    content_type = models.ForeignKey(ContentType, models.CASCADE, verbose_name="content type")
+    codename = models.CharField("codename", max_length=100)
+
+    class Meta:
+        verbose_name = "permission"
+        verbose_name_plural = "permissions"
+        unique_together = [["content_type", "codename"]]
```

Comment:
Codnames need to either be unique by themselves or this model needs to have some kind of method for creating a unique permission key using the codename and the object_type. These codenames will have to be used to communicate permissions between systems, and we can't have any ambiguity about what permission a particular codename corresponds to.

I would really prefer the latter, with some unique identifier that includes the ansible service type (awx, galaxy, etc), model name and codename (though the AAP service type isn't strictly necessary here). Maybe we have an "action" field and generate the codename field from the content type name and action? `codename = f"{action}_{content_type.model}"` or something like that. Even better if it's formatted like `{content_type.model}.{action}` for readability :)

Another option would be a UUID field, but that's not readable at all.

Reply:
reply in mainline discussion / another issue